sdk/typescript/experimental/src/sui-client.ts

```diff
@@ -0,0 +1,271 @@
+     * A URL to send transactions to.
+     * To be removed once Full Node starts accepting txs.
+     */
+    gatewayUrl: string,
+    /**
+     * A URL to query data from.
+     */
+    fullNodeUrl: string,
+    /**
+     * A gas to use when sending transactions.
+     * Version is automatically upated from tx response.
+     */
+    protected gasRef: SuiObjectRef | null = null
+  ) {
+    this.rpc = new RpcClient(gatewayUrl, fullNodeUrl);
+  }
+
+  /**
+   * Fetch a local ref if it exists, return null otherwise
+   * @param id ID of the object
+   * @param force Whether to ignore cached version
+   */
+  public async ref(id: string, force: boolean = false): Promise<SuiObjectRef> {
+    if (force) {
+      return this.fetchObject(id);
+    }
+
+    return this.objects.has(id)
+      ? Promise.resolve(this.objects.get(id)!)
+      : this.fetchObject(id);
+  }
+
+  /**
+   * Get owned objects' references grouped by a type.
+   * @param address Address of an account to fetch objects for
+   * @returns Object where keys are types and values are arrays of references
+   */
+  public async myObjectsByType(
+    address: string
+  ): Promise<{ [key: string]: SuiObjectRef[] }> {
+    let objects = await this.rpc.myObjects(address);
+    let byType: { [key: string]: SuiObjectRef[] } = {};
+    for (let obj of objects) {
+      obj.type in byType
+        ? byType[obj.type].push(obj)
+        : (byType[obj.type] = [obj]);
+
+      this.objects.set(obj.objectId, {
+        objectId: obj.objectId,
+        version: obj.version,
+        digest: obj.digest,
+      });
+    }
+
+    return byType;
+  }
+
+  /**
+   * Fetch an object from the network
+   * @param id ID of the object
+   */
+  public async fetchObject(id: string): Promise<SuiObjectRef> {
+    let ref = await this.rpc.fetchObjRef(id);
+    this.objects.set(ref.objectId, ref);
+    return ref;
+  }
+
+  /**
+   * Balance of the current gas object used
+   */
+  public async gasBalance(): Promise<number> {
+    const gas = this.gasOrFail();
+
+    return this.rpc.fetchObj(gas.objectId).then((res) => {
+      if (res.status === "Exists") {
+        return res.details.data.fields.balance;
+      } else {
+        throw new Error(
+          `Gas Object not found: ${gas.objectId}! \n${JSON.stringify(res)}`
+        );
+      }
+    });
+  }
+
+  /**
+   * Update gas object used for transactions; if it exists - use, if not -
+   * query it from the network.
+   * @param gasId ID of the Gas object to use
+   */
+  public async setGas(gasId: string): Promise<void> {
+    if (this.objects.has(gasId)) {
+      this.gasRef = this.objects.get(gasId)!;
+    } else {
+      this.gasRef = await this.rpc.fetchObjRef(gasId);
+    }
+  }
+
+  /**
+   * Send the transaction to the network and update local versions.
+   * Currently only supports `MoveCallTx` and a `PublishTx`
+   */
+  public async send(
+    keypair: KeyPair,
+    tx: MoveCallTx | PublishTx,
+    gasBudget = 5000,
+    bcsBufferSize = 4096
```

Comment:
@damirka , how do the caller estimate this number? Is there way to derive it from the TxData?

Reply:
I assume there is, yes. The buffer is then cut to match the data but die to Uint8Array allocation specifics, I did not manage to implement it. 1-4KB is more than enough for most purposes (except publishing).

Reply:
Nice catch! We need to come up with a solution. I’ll look into it today.